File: src/cviz/timeline/Trigger.java
package cviz.timeline;

import cviz.timeline.commands.ICommand;
import cviz.timeline.commands.LoopCommand;

import java.util.LinkedList;
import java.util.List;

public class Trigger {
	private final TriggerType type;
	private final long targetFrame;
	private final int layerId;
	private final boolean loop;
	private boolean waited = false;
	
	private final LinkedList<ICommand> commands = new LinkedList<>();

	public static Trigger CreateCue(){
		return new Trigger(TriggerType.CUE, -1, -1, false);
	}
	public static Trigger CreateImmediate(){
		return new Trigger(TriggerType.IMMEDIATE, -1, -1, false);
	}
	public static Trigger CreateFrame(int layerId, long targetFrame){
		return new Trigger(TriggerType.FRAME, layerId, targetFrame, false);
	}
	public static Trigger CreateEnd(int layerId){
		return new Trigger(TriggerType.END, layerId, -1, false);
	}
	public static Trigger CreateLoop(int layerId){
		Trigger t = new Trigger(TriggerType.END, layerId, -1, true);
		t.addCommand(new LoopCommand(layerId));
		return t;
	}

	private Trigger(TriggerType type, int layerId, long targetFrame, boolean loop) {
		this.type = type;
		this.targetFrame = targetFrame;
		this.layerId = layerId;
		this.loop = loop;
	}
	
	protected void addCommand(ICommand c) {
		commands.add(c);
	}

    public List<ICommand> getCommands(){
        return commands;
    }

	public TriggerType getType() {
		return type;
	}
	
	public long getTargetFrame() {
		return targetFrame;
	}
	
	public int getLayerId() {
		return layerId;
	}
	
	public void setWaited() {
		waited = true;
	}
	
	public boolean hasWaited() {
		return waited;
	}
	
	public boolean isLoop() {
		return loop;
	}

	public String toString() {
		return "Trigger: " + type + " " + layerId + " " + targetFrame + " loop: " + loop;
	}
}

File: CViz/Timeline/Trigger.cs
using System.Collections.Immutable;
using CViz.Timeline.Command;

namespace CViz.Timeline
{
    class Trigger
    {
        public TriggerType Type { get; }
        public string Name { get; }
        public int LayerId { get; }
        public long TargetFrame { get; }
        public bool Loop { get; }
        public bool Waited { get; private set; }

        public ImmutableList<CommandBase> Commands { get; }

        public static Trigger CreateCue(string name) => new Trigger(TriggerType.Cue, name);
        public static Trigger CreateSetup() => new Trigger(TriggerType.Setup, "Setup");
        public static Trigger CreateEnd(int layerId) => new Trigger(TriggerType.End, string.Format("{0} End", layerId));

        public static Trigger CreateFrame(int layerId, long targetFrame)
        {
            return new Trigger(TriggerType.Frame, string.Format("{0} F{1}", layerId, targetFrame), layerId, targetFrame);
        }

        public static Trigger CreateLoop(int layerId)
        {
            Trigger t = new Trigger(TriggerType.End, string.Format("{0} Loop", layerId), layerId, loop: true);
            return t.WithCommand(new LoopCommand(layerId));
        }

        private Trigger(TriggerType type, string name, int layerId = -1, long targetFrame = -1, bool loop = false, ImmutableList<CommandBase> commands=null)
        {
            Type = type;
            Name = name;
            LayerId = layerId;
            TargetFrame = targetFrame;
            Loop = loop;

            Commands = commands ?? ImmutableList<CommandBase>.Empty;
        }

        public Trigger WithCommand(CommandBase cmd)
        {
            ImmutableList<CommandBase> cmds = Commands.Add(cmd);
            return new Trigger(Type, Name, LayerId, TargetFrame, Loop, cmds);
        }

        public void SetWaited()
        {
            Waited = true;
        }

        public override string ToString()
        {
            return string.Format("Trigger: ({0}) {1} {2} {3} Loop:{4}", Name, Type, LayerId, TargetFrame, Loop);
        }
    }
}
